### ISSUE TITLE
Out-of-memory issue

### ISSUE THREAD

**Issue by liangwang0734** [ornladios/ADIOS2#2892]
**If I know my per-processor memory upper limit, how can I tell ADIOS2 to do the actual writing to disk before memory runs out?**

My task is to read a chunk of data from many small files (~500k) and save them in a bp file folder (after processing).
- Each processor handles a small portion of the files.
- `Sync` mode is used in `Put` so that I can reuse my own buffer for reading from the small input files.
- In my test, the data size to write is about 4T, the total memory available across all nodes might be around 6-7T.
- I understand ADIOS2 uses an internal buffer for optimized output; maybe what I need to do is to set the buffer size?

=====================
##### My original code using `fstream` and questions,

Later I switched to use the lower-level APIs (io, engine, variable, etc.). But the memory issue persists.

```cxx
adios2::fstream oStream(output_filename, adios2::fstream::out, MPI_COMM_WORLD);
for (int i=0; i<ifile1; ++ i) {
  // read data from the ith file into buf, and shift the start index in the global array
  oStream.write("output", buf.data(), shape, start, count);
}
oStream.close();
```

My code would be fast in the iteration but dies from out-of-memory error at wrapping up.
Overriding the `endStep` option in `oStream.write` seems to solve the problem, though the iteration becomes much slower:
```cxx
oStream.write("output", buf.data(), shape, start, count, true);
```

So my questions are:
- By default, does the `oStream.write` call defer the IO or not? The document describes the last parameter as "endStep: similar to std::endStep, end current step and flush (default). Use adios2::endStep if true", which seems to indicate the default is to flush. But this seems to contract what I encountered.
- Is there a portable way to let ADIOS2 choose whether and when to flush to avoid memory error?

**Comment by williamfgc** [comment]
@liangwang0734 this is the high level C++ API mostly used for testing and quick data access as `write` is sync. For HPC application development use the full APIs. Docs are [here](https://adios2.readthedocs.io/en/latest/api_full/api_full.html#full-apis). Let us know if you have any questions.

**Comment by liangwang0734** [comment]
Thank you, @williamfgc 

I turned to use the lower API, but had trouble reproducing the correct results from `fstream` implementation.
- I surely wrongly understood some very basic things here, either in data order or the flushing time.
- I did not use `engine.BeginStep()/EndStep()` pairs because this [example](https://adios2-examples.readthedocs.io/en/latest/concepts/anatomy.html#file-only-input) says it is not needed for file-only operations (I'm not doing code coupling).
  - If I call them around `engine.Put` the results are totally garbage.

```cxx
// global array shape is {Nz, Ny, Nx}
// local block shape is {nz, ny, nx}
// each processor handles blocks with index iblock_beg <= i < iblock_end

// method 1, using fstream
adios2::fstream oStream(output_filename_1, adios2::fstream::out,
                      	MPI_COMM_WORLD);

// method 2, using lower API
adios2::ADIOS adios(MPI_COMM_WORLD);
adios2::IO io = adios.DeclareIO("MergedDataAthenaVtk3D");
adios2::Engine engine = io.Open(output_filename_2,adios2::Mode::Write);
adios2::Variable<float> var = io.DefineVariable<float>(field_name, {Nz, Ny, Nx});

for (int i=iblock_beg; i<iblock_en; ++i) {
	// get/read data of the ith block, process it, and save it in a local buf

	// compute iz0, iy0, ix0 from block index i

	// method 1, using fstream
	const adios2::Dims shape{Nz, Ny, Nx};
	const adios2::Dims count{nz, ny, nx};
	const adios2::Dims start{iz0, iy0, ix0};
	oStream.write(field_name, buf.data(), shape, start, count);

	// method 2, using lower API
	var.SetSelection({{iz0, iy0, ix0}, {nz, ny, nx}});
	engine.Put(var, buf.data());
}

engine.close(); // results 
oStream.close();
```
Method 1, using `fstream`, correct
![correct](https://user-images.githubusercontent.com/1000547/135724574-f51f64e0-307a-4f1c-940f-f293b51f0086.png)


Method 2, using lower APIs, wrong
![wrong-method-2](https://user-images.githubusercontent.com/1000547/135724578-ab6acbfd-1fb0-4acf-b749-6492fdae4a2c.png)

**Comment by williamfgc** [comment]
> I did not use engine.BeginStep()/EndStep() pairs because this example says it is not needed for file-only operations (I'm not doing code coupling).

@liangwang0734 the referred docs are in the context of file input (reading) random access. My suggestion is to first understand the concepts and motivation for streaming and file random access APIs. adios2 introduce concepts like step-based and deferred based  I/O. I'd stick with a single API (full C++) before the high level APIs as those are sync and not meant for performance (they build on top of core components). The [paper](https://www.sciencedirect.com/science/article/pii/S2352711019302560) and the [docs](https://adios2.readthedocs.io/en/latest/) might be a good starting point, examples illustrate these concepts. Hope it helps, let us know if you have any questions.

**Comment by liangwang0734** [comment]
@williamfgc 
Yes, I need to understand the concepts better.

Right now, I can run the program to get correct results using the full API (io, engine, variable).

But the memory issue remains, which led me to suspect that I need to know how to set the size of the internal buffer, or how to force flush properly. Or how to apply the step concept to my block iteration
 - each iteration fills some blocks into the global array
 - the more steps, the more blocks filled;
 - the global array is never overwritten

**Comment by williamfgc** [comment]
@liangwang0734 let's take a step back and try to understand your use case I/O pattern. Are you dumping the same variables at every step? If so, BeginStep/EndStep is what is needed. APIs are contracts, and the motivation for their existence is particular use cases. Once your need is established, we can dig into the proper API example.

User Memory from variable data in adios2 is copied to a buffer by default at PerformPuts, EndStep or Close. There is the Variable Span functionality that exposes adios2 memory buffer for memory-bound applications that can reuse this memory....it ends up being several concepts, semantics, options that need to be clear before using a particular API. Hope it helps.

**Comment by pnorbert** [comment]
Hi @liangwang0734. Forget about the Flush() function, it does not do what you want to do. The BP4 file engine uses it to write to disk **multiple Steps** when you have a large enough buffer pre-allocated to contain multiple steps. Even in that case (which is definitely not yours), it's better to let the engine deal with the flush rather than the user because it only works properly between EndStep..BeginStep, i.e. with whole steps buffered.

So, currently in ADIOS there is no way to generate a global array so big that it does not fit into the memory. The one (very big assumption or constraint) in the ADIOS API is that one step does fit into the memory entirely. You need to use more compute nodes enough to buffer all the local blocks at once. And then, if you use aggregation, you need even more memory (twice as much at least). 

In the future, the new BP5 engine will allow to construct such a big array by introducing a new function that will work like what you expected from a flush function. But right now, you need to run on more nodes.

**Comment by liangwang0734** [comment]
@pnorbert Thank you. 

My task is to read a block of data from many small files (~500k) and save them in a bp file folder.

- Each processor handles a small portion of the files.
  - For example, rank 0 gets block 0-249, rank 1 gets block 250-499
- The locations of the blocks in the global array is not consecutive in terms of the block ID, but can be computed from the block ID.
- Each block actually contains multiple physical variables. But for simplicity, we may just extract one variable.
- In my test, the data size to write is about 4 TB, the total memory available across all nodes might be around 6-7 TB.
- Presently, two ways work
 ```
// method 1: sync mode
for each block on this process {
  // read data from the block
  // set start index: iz0, iy0, ix0
  var.SetSelection({{iz0, iy0, ix0}, {nz, ny, nx}});
  engine.Put(var, buf.data(), adios2::Mode::Sync);
 }

// method 2: deferred mode + PerformPuts
for each block on this process {
  read data from the block
  // set start index: iz0, iy0, ix0
  var.SetSelection({{iz0, iy0, ix0}, {nz, ny, nx}});
  engine.Put(var, buf.data(), adios2::Mode::Sync);
  engine.PerformPuts();
}
```
- I did try `BeginStep/EndStep` or `Flush`. But it looks like they save each step separately instead of accumulatively.

**Comment by liangwang0734** [comment]
> So, currently in ADIOS there is no way to generate a global array so big that it does not fit into the memory. The one (very big assumption or constraint) in the ADIOS API is that one step does fit into the memory entirely. You need to use more compute nodes enough to buffer all the local blocks at once. And then, if you use aggregation, you need even more memory (twice as much at least).
> 
> In the future, the new BP5 engine will allow to construct such a big array by introducing a new function that will work like what you expected from a flush function. But right now, you need to run on more nodes.

@pnorbert I see. This greatly explains things for me. In my case, the amount of data to write exceeds the total memory available.

Is there any possibility to trigger data dump (to disk) and make the internal buffer available for next dump? This was possible for, for example, HDF5 (as I can manually call their `flush` when needed) for any size of dataset.

I'm looking forward to bp5.

**Comment by pnorbert** [comment]
Note that in the BP format, those 500k blocks of data will be 500k separate blocks in the data files. Even though you see Nx*Ny*Nz array as the global shape, reading the data back will take ages, depending on the read pattern. If you have thousands of readers, if you read a 2D slice, these large number of blocks may be beneficial, but forget about processing it on a serial process. It will still be much faster than opening 500k files, but beating that will not make you satisfied. 

In other words, ADIOS does not do data reorganization for you. You could use `adios2_reorganize` later on N process to read back and then output the entire variable in N blocks instead of 500k but it will take a long time to run, and will run out of memory anyway. 

What is your experience using HDF5 with chunking? Did that finish in usable time for creating this dataset, and for reading it back?

**Comment by liangwang0734** [comment]
> What is your experience using HDF5 with chunking? Did that finish in usable time for creating this dataset, and for reading it back?

I need some library that can access data as a regular `Nz * Ny * Nx` array. So HDF5 and ADIOS2 would work for me.

Creating the single-file HDF5 was a pain. My implementation of parallel writing (in mpi4py and h5py-parallel) didn't seem to be much faster than serial io. I ended up using serial h5py and it usually takes 2 days to create a 2 TB chunked hdf5 data, that was OK for me, but surely wasn't nice (and that's why I tried ADIOS5, together with the difficult C API of hdf5).

I had used chunked HDF5 data for parallel FFT (of size 10k^3 but I don't remember the chunk size, maybe 100^3?) over a few thousands of cores (maybe around 5k cores?). The reading time was between 10 mins to 15 mins.

BTW, I also tried numpy mmap, but it simply cannot map large arrays.

**Comment by liangwang0734** [comment]
> Note that in the BP format, those 500k blocks of data will be 500k separate blocks in the data files. Even though you see Nx_Ny_Nz array as the global shape, reading the data back will take ages, depending on the read pattern.

Is there a way to give hints to find a better block size, like 2^3 times bigger blocks, which would reduce the block size by 8 times?

**Comment by williamfgc** [comment]
@liangwang0734 adios is MPI based. It benefits from distributed memory and a parallel file system. That's the main use of global arrays. As long as you have enough memory it should work on one node.

> Is there a way to give hints to find a better block size, like 2^3 times bigger blocks, which would reduce the block size by 8 times?

That depends on you access pattern. The more often you access memory contiguous blocks (in file and in memory when slicing) the better, as there will be fewer calls to expensive functions like read and memcpy under the hood. Ideally, the read pattern should be known.

**Comment by pnorbert** [comment]
Re. out of memory: for BP4 your approach could be this:
- let adios pre-allocate a big enough buffer to contain all your data from the process + a few megabytes for metadata
- use Put(Sync) to make adios buffer the data immediately
- free your user buffer immediately to avoid having two copies build up
- set NumAggregators=_N_ where _N_ is the number of processes, which you said was 2000

This way you will only  need a little above the 4TB data size for building the entire array in memory and not use extra copies of it for aggregation. So you should be able to run the program on a number of nodes with aggregate memory of 4.2TB or so, maybe a bit more. 

You would end up with 2000 files in the .bp directory if you run on 2000 processes, which may be too many on the Pleiades system's file system. Since you are constrained by the file system in every step, opening files from 2000 processes, reading data at once from 2000 processes, then writing to 2000 files (which would be a good number for Summit's giant file system), I feel using less processes could actually improve the overall runtime. What was your experience with HDF5 if you had the patience to evaluate this?

With this scheme then, BP5 later would let you do aggregation to less number of files without extra copies, as well as save the trouble of preallocating a big enough but not too big ADIOS buffer (and hopefully, improve performance by better handling the metadata for 500k blocks per variable).

**Comment by pnorbert** [comment]
> 
> ```
> // method 1: sync mode
> for each block on this process {
>    for each variable {
>    // get hold of the correct buf and adios2::Variable var
>    // read data from the block into buf
>    // set start index: iz0, iy0, ix0
>    var.SetSelection({{iz0, iy0, ix0}, {nz, ny, nx}});
>    engine.Put(var, buf.data(), adios2::Mode::Sync);
>  }
> }
> 
> // method 2: deferred mode + PerformPuts
> for each block on this process {
>  for each variable {
>    // get hold of the correct buf and adios2::Variable var
>    // read data from the block into buf
>    // set start index: iz0, iy0, ix0
>    var.SetSelection({{iz0, iy0, ix0}, {nz, ny, nx}});
>    engine.Put(var, buf.data(), adios2::Mode::Deferred);
>  }
>  engine.PerformPuts();
> }
> ```

So just  a note, with your second approach there is a point (at PerformPuts), where you have two copies of the entire array in memory: all your read-in chunks passed in Put(), and the ADIOS buffer being built during PerformPuts.

**Comment by pnorbert** [comment]
> > Note that in the BP format, those 500k blocks of data will be 500k separate blocks in the data files. Even though you see Nx_Ny_Nz array as the global shape, reading the data back will take ages, depending on the read pattern.
> 
> Is there a way to give hints to find a better block size, like 2^3 times bigger blocks, which would reduce the block size by 8 times?

As I said, ADIOS does not do data reorganization, so it will not merge the blocks in any pattern. I.e., it does not have something akin to HDF5's chunking concept. And if it had, a general solution in the I/O library that works for your use case would be as performant as that one. It takes a lot of time for many processes shuffle all that data around to create the chunks.

The best approach would be for you to merge the blocks yourself before passing to the I/O library. You mentioned you can calculate the position of the blocks from the block ID. So, can you first sort them so that you can give a series of block IDs to a process that form a joined giant block in the global array (the easiest would be to merge them along the slowest dimension)?

Then, you could read in the blocks one by one into a larger contiguous buffer and thus merge a bunch of them along the slow dimension. Then pass this big block to ADIOS/HDF5. 

I assumed here that nx, ny, nz are uniform size for all blocks. And if you cannot do this from the block number but you have to open each file and read in the ID, you could have two programs, the first only creating this ordering. Assuming it is not the serial opening of 500k files that takes most of the 2 days of runtime, this could lead to a big improvement in performance.
